stan/math/prim/scal/meta/enable_if_generics.hpp

```diff
@@ -0,0 +1,810 @@
+#ifndef STAN_MATH_PRIM_SCAL_META_ENABLE_IF_GENERICS_HPP
+#define STAN_MATH_PRIM_SCAL_META_ENABLE_IF_GENERICS_HPP
+
+#include <stan/math/prim/scal/meta/conjunction.hpp>
+#include <stan/math/prim/scal/meta/disjunction.hpp>
+#include <stan/math/prim/scal/meta/scalar_type.hpp>
+#include <stan/math/prim/scal/meta/is_var.hpp>
+#include <stan/math/prim/scal/meta/is_vector.hpp>
+#include <stan/math/prim/scal/meta/is_vector_like.hpp>
+#include <stan/math/prim/scal/meta/is_eigen.hpp>
+#include <stan/math/prim/scal/meta/is_fvar.hpp>
+#include <stan/math/prim/scal/meta/is_var_or_arithmetic.hpp>
+
+#include <type_traits>
+
+namespace stan {
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution. This method performs a negation of
+ * the input boolean.
+ */
+template <typename Check>
+using disable_if = typename std::enable_if_t<!Check::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if all conditions are true and otherwise fails.
+ */
+template <class... Checks>
+using enable_if_all = std::enable_if_t<math::conjunction<Checks...>::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if any of the conditions are true and otherwise fails.
+ */
+template <class... Checks>
+using enable_if_any = std::enable_if_t<math::disjunction<Checks...>::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if all of the conditions are false.
+ */
+template <class... Checks>
+using disable_if_all = std::enable_if_t<!math::disjunction<Checks...>::value>;
+
+/**
+ * Template metaprogram for deducing if a method should be remove a
+ * a function from overload resolution.
+ *
+ * Returns a type void if any of the conditions are false.
+ */
+template <class... Checks>
+using disable_if_any = std::enable_if_t<!math::conjunction<Checks...>::value>;
```

Comment:
All the stuff below here are just using the alias's above for specializations so we can write stuff like `enable_if_eigen_var<T>` etc.